### ISSUE TITLE
Inserting images can lead to "reformat" warning

### ISSUE THREAD

**Issue by indigoxela** [backdrop-contrib/ckeditor5#86]
To reproduce:

- Create a new node
- Create text (paragraph)
- Insert a new image (don't do anything else, just insert after that text)
- Save
- Open the node form again

I get the warning "Activating CKEditor 5 will reformat the content of this field. ...", and in fact, the img tag isn't wrapped in a paragraph, instead there's an empty newline between "p" and "img", which gets removed after toggling the editor on.

Another downcast/upcast riddle?

**Comment by quicksketch** [comment]
Thanks for filing this. I noticed I'm getting the reformat warning more than I should be as well.

**Comment by quicksketch** [comment]
It seems like this might be fixed as a side-effect of the PR at https://github.com/backdrop-contrib/ckeditor5/pull/88.

**Comment by indigoxela** [comment]
> It seems like this might be fixed as a side-effect of the PR at https://github.com/backdrop-contrib/ckeditor5/pull/88.

Hm, I'm not sure, I still get the warning, when adding an image after text. And the empty newline's still there.

I also tried to add an image before the text, inside the paragraph, then there's no empty newline after img, but the warning still pops up and the editor refuses to initialize.

In this case (img inside p, before text) the empty newline gets appended after the p tag, and gets removed when initializing CKE5.

**Comment by quicksketch** [comment]
The problem might actually the attribute order. After inserting an image, it is this order:

```
<img alt="" src="/files/field/image/card1-layout.png" width="800" height="533">
```

But after saving and re-editing, the order is now:

```
<img src="/files/field/image/card1-layout_7.png" alt="" width="800" height="533">
```

I see two possible approach here: we make inserting images use the preferred attribute order with src first in the Image plugin, or we alphabetically order the attributes (or make some other kind of predetermined order) in the HTML writer plugin.

I don't really like alphabetical attributes, especially with images, since you end up with attributes in the order: `height`, `src`, `width`, and I'd really like width and height next to each other... but I'm not sure how else we're going to be consistent about attribute order if the browser logic is going to be unpredictable.

**Comment by quicksketch** [comment]
@indigoxela I made a proof of concept PR at https://github.com/backdrop-contrib/ckeditor5/pull/89 that I think fixes this issue. Can you verify?

A better option that would take more work but would avoid the alphabetical attributes would be if the `Backdrop.ckeditor5.formatHtml()` function could _optionally_ sort the attributes. Then we could sort the attributes for the purposes of comparison, but preserve the attribute order when we save the data. But right now that code is only responsible for creating new lines and indentation, it doesn't actually affect the markup in any other way, so expanding it to do something like that would take a lot more adjustment to the code.

**Comment by indigoxela** [comment]
> The problem might actually the attribute order. 

Yes, it's actually the order! :tada: 

I can confirm, that the POC PR fixes the warning problem.

>  we could sort the attributes for the purposes of comparison, but preserve the attribute order when we save the data

That's an option, but... it means some hidden logic, that might be hard to debug, if we ever run into problems with it. Figuring out that something has to do with a temporary change in the conversion process might be difficult (or maybe not... I'm not sure).

Re whitespace changes: this still happens, but doesn't seem to have any side effects. Even if the "Convert line breaks into HTML" filter is turned on. Or can that be relevant for content display at some point?
Currently there's this subtle difference in markup between "Source mode" in the editor and when switching to "Raw HTML" filter. Maybe we can just ignore that.

**Comment by indigoxela** [comment]
Another tiny thing: self-closing tags in CKE4 vs. CKE5:

- Create a node with CKE4, insert an image (only)
- Save node
- Open node form again
- Switch to CKE5
- You get the warning

That's subtle, it's only about how these tags are written:

4:
```
<img alt="" height="533" src="/files/field/image/card1-layout.png" width="800" />
```

5:
```
<img alt="" height="533" src="/files/field/image/card1-layout.png" width="800">
```

Note the XML-style slash added by 4, while 5 removes it. Toggling again adds or removes the slash, so the warning will again pop up.

**Comment by quicksketch** [comment]
Thanks @indigoxela again for the analysis. So we have the potential differences that should be ignored when comparing existing content to content modified by CKEditor:

- Line indentation.
- Other whitespace like new lines.
- Attribute order.
- Self-closing tags.

**Comment by quicksketch** [comment]
I filed a new PR at https://github.com/backdrop-contrib/ckeditor5/pull/91 that might solve this problem, but it's a little finicky so I'm not sure if it's going to work as well as I would want it to.

I moved the `BackdropHtmlBuilder` class out of the `backdrop-html-engine` CKEditor plugin so that it can be used separately. Now the default value of the field is run both through `BackdropHtmlBuilder` (to standardize self-closing tags and attribute order) and then `BackdropHtmlFormatter` (to make white space consistent). Then it's compared with the value that comes out of CKEditor. I tried to combine both of these classes together so it's done all at once, but it was both difficult to write and I found that CKEditor doesn't do great when combining already-formatted HTML with the Source button. It reformats the HTML a second time when showing source mode, resulting in double-indentation and extra new lines. So we need to have the processes of attribute sanitation be separate from white space formatting.

But this approach is much less new code; it's just rearranging where the current code is. I wish they could be better integrated but this seems to work.

**Comment by indigoxela** [comment]
This new approach (#91) seems to work like a charm! :tada:

- Created some content (img and hr tag) in CKE4
- Switched to 5, no warning
- Switched back to 4 no problem (toggling changes, how self-closing tags are written and source formatting)
- Added caption with markup (bold)
- More 4/5 toggling, source mode checks
- Saved with 5 - markup in caption displays correctly
- Opened again, still all good :+1: 

Works for me!

**Comment by quicksketch** [comment]
Wonderful, thank you for testing @indigoxela! I merged #91.
